### PR TITLE
Add Diagnostic Dumping Functionality

### PR DESCRIPTION
**Pull request by mdaneri**
#### Description

This pull request introduces a new function, **Invoke-PodeDump**, to capture detailed diagnostic information during critical failures or fatal exceptions within Pode.

#### Key Features

1. **Configurable Dump Formats**:
   - Supports formats such as JSON, YAML, CLIXML, Plain Text, and Binary.

2. **Pipeline Support**:
   - The `ErrorRecord` parameter accepts pipeline input, simplifying integration with try-catch blocks.

3. **Dynamic Overrides**:
   - Supports runtime overrides for `Path` and `Format`.
   - Default configuration options can be set in `server.psd1`.

4. **Optional Application Halt**:
   - Includes a `-Halt` switch to terminate the application after saving the diagnostic dump.

#### Configuration in `server.psd1`

```powershell
@{
    Server = @{
        Debug = @{
            Dump = @{
                Enable      = $true
                Format      = 'Yaml'
                Path        = './Dump'
                MaxDepth    = 5
            }
        }
    }
}
```

- **Enable**: Enable or disable memory dump functionality.
- **Format**: Set the default output format.
- **Path**: Specify the directory for saving dump files.
- **MaxDepth**: Define the maximum depth of information collection.

#### Example Usage

1. **Diagnostic Dump with Halt**:
   ```powershell
   try {
       throw [System.Exception] "Simulated exception"
   }
   catch {
       $_ | Invoke-PodeDump -Format 'json' -Halt
   }
   ```

2. **Custom Configuration Overrides**:
   ```powershell
   try {
       throw [System.Exception] "Simulated error"
   }
   catch {
       $_ | Invoke-PodeDump -Format 'yaml' -Path 'C:\CustomDumps'
   }
   ```

#### Notes

- Automatically creates directories for the specified dump path if they do not exist.
- Enhances Pode's error handling capabilities by providing detailed diagnostics for debugging.

### CODE REVIEW

File: src/Private/Dump.ps1 [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+    # Embed C# code to handle the DebuggerStop event
+    Add-Type @'
+using System;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+using System.Collections.ObjectModel;
+
+namespace Pode.Embedded
+{
+    public class DebuggerHandler
```
Comment (Badgerati) [inline src/Private/Dump.ps1:503]: I started testing and hit the first issue with this class, since all of the properties and methods are static it causes subsequent calls after the first to become contaminated - for example the first runspace sets `isEventTriggered` to `true`, but then it's true for all other runspaces even though it's actually false.

I fixed the C# code to be the following further below - including updating how the Dump works to below that. BUT, that's when I hit the first major snag, with the class fixed and using `Web-Dump.ps1` example, all of the "waiting for debugs" timeout after 60s - and they will for basically ever until several appropriate requests hit the server as Pode is waiting for inbound requests.

image

Specifically it's lines like below which hang waiting, in `PodeServer.ps1`:
```powershell
$context = (Wait-PodeTask -Task $Listener.GetContextAsync($PodeContext.Tokens.Cancellation.Token))
```

I'm trying to think of ways around this, and one idea I had was to give each Runspace its own CancellationToken and this is used on `.GetContextAsync` calls. When we try to attach the debugger to a runspace, we `Enable-RunspaceDebug` and then Cancel that runspace's CancellationToken, and then renew it for the next loop run - that "should" stop the timeouts.

When the server stops/restarts, then we need to Cancel all the runspace tokens as well.

### Fix DebuggerHandler
```c#
using System;
using System.Management.Automation;
using System.Management.Automation.Runspaces;
using System.Collections.ObjectModel;

namespace Pode.Embedded
{
    public class DebuggerHandler : IDisposable
    {
        // Collection to store variables collected during the debugging session
        public PSDataCollection<PSObject> Variables { get; private set; } = new PSDataCollection<PSObject>();

        // Event handler for the DebuggerStop event
        private EventHandler<DebuggerStopEventArgs> DebuggerStopHandler;

        // Flag to indicate whether the DebuggerStop event has been triggered
        public bool IsEventTriggered { get; private set; } = false;

        // Flag to control whether variables should be collected during the DebuggerStop event
        private bool CollectVariables = true;

        // Runspace object to store the runspace that the debugger is attached to
        private Runspace Runspace;

        public DebuggerHandler(Runspace runspace, bool collectVariables = true)
        {
            // Set the collection flag and Runspace object
            CollectVariables = collectVariables;
            Runspace = runspace;

            // Initialize the event handler with the OnDebuggerStop method
            DebuggerStopHandler = new EventHandler<DebuggerStopEventArgs>(OnDebuggerStop);
            Runspace.Debugger.DebuggerStop += DebuggerStopHandler;
        }

        // Method to detach the DebuggerStop event handler from the runspace's debugger, and general clean-up
        public void Dispose()
        {
            IsEventTriggered = false;

            // Remove the event handler to prevent further event handling
            if (DebuggerStopHandler != default(EventHandler<DebuggerStopEventArgs>))
            {
                Runspace.Debugger.DebuggerStop -= DebuggerStopHandler;
                DebuggerStopHandler = null;
            }

            // Clean-up variables
            Runspace = default(Runspace);
            Variables.Clear();

            // Garbage collection
            GC.SuppressFinalize(this);
        }

        // Event handler method that gets called when the debugger stops
        private void OnDebuggerStop(object sender, DebuggerStopEventArgs args)
        {
            // Set the eventTriggered flag to true
            IsEventTriggered = true;

            // Cast the sender to a Debugger object
            var debugger = sender as Debugger;
            if (debugger == default(Debugger))
            {
                return;
            }

            // Enable step mode to allow for command execution during the debug stop
            debugger.SetDebuggerStepMode(true);

            // Collect variables or hang the debugger
            var command = new PSCommand();
            command.AddCommand(CollectVariables
                ? "Get-PodeDumpScopedVariable"
                : "while($PodeContext.Server.Suspended) { Start-Sleep -Milliseconds 500 }");

            // Execute the command within the debugger
            var outputCollection = new PSDataCollection<PSObject>();
            debugger.ProcessCommand(command, outputCollection);

            // Add results to the variables collection if collecting variables
            if (CollectVariables)
            {
                foreach (var output in outputCollection)
                {
                    Variables.Add(output);
                }
            }
            else
            {
                // Ensure the debugger remains ready for further interaction
                debugger.SetDebuggerStepMode(true);
            }
        }
    }
}
```

### Updated Get-PodeRunspaceVariablesViaDebugger
```powershell
function Get-PodeRunspaceVariablesViaDebugger {
    param (
        [Parameter(Mandatory)]
        [System.Management.Automation.Runspaces.Runspace]$Runspace,

        [Parameter()]
        [int]$Timeout = 60
    )

    # Initialize variables collection
    $variables = @()
    try {

        # Attach the debugger and break all
        $debugger = [Pode.Embedded.DebuggerHandler]::new($Runspace)
        Enable-RunspaceDebug -BreakAll -Runspace $Runspace

        # Wait for the event to be triggered or timeout
        $startTime = [DateTime]::UtcNow
        Write-PodeHost "Waiting for $($Runspace.Name) to enter in debug ." -NoNewLine

        while (!$debugger.IsEventTriggered) {
            Start-Sleep -Milliseconds 500
            Write-PodeHost '.' -NoNewLine

            if (([DateTime]::UtcNow - $startTime).TotalSeconds -ge $Timeout) {
                Write-PodeHost "Failed (Timeout reached after $Timeout seconds.)"
                return @{}
            }
        }

        Write-PodeHost 'Done'

        # Retrieve and output the collected variables from the embedded C# code
        $variables = $debugger.Variables
    }
    catch {
        # Log the error details using Write-PodeErrorLog.
        # This ensures that any exceptions thrown during the execution are logged appropriately.
        $_ | Write-PodeErrorLog
    }
    finally {
        # Detach the debugger from the runspace to clean up resources and prevent any lingering event handlers.
        if ($null -ne $debugger) {
            $debugger.Dispose()
        }

        # Disable debugging for the runspace. This ensures that the runspace returns to its normal execution state.
        Disable-RunspaceDebug -Runspace $Runspace
    }

    return $variables[0]
}
```

**Review by Badgerati — CHANGES_REQUESTED** [review]
@mdaneri There's a fair few comments, but the most important ones are for the DebuggerHandler class, so I would recommend reading those first.

File: docs/Tutorials/ServerOperation/Restarting/Overview.md [hunk 2]
```diff
@@ -9,3 +9,5 @@ There are 4 ways to restart a running Pode server:
 4. [`Restart-PodeServer`](../../../Functions/Core/Restart-PodeServer): This function lets you manually restart Pode from within the server.
 
 When the server restarts, it will re-invoke the `-ScriptBlock` supplied to the [`Start-PodeServer`](../../../Functions/Core/Start-PodeServer) function. This means the best approach to reload new modules/scripts it to dot-source/[`Use-PodeScript`](../../../Functions/Utilities/Use-PodeScript) your scripts into your server, as any changes to the main `scriptblock` will **not** take place.
+
+Certainly! Here’s an explanation on how to achieve suspending and resuming a Pode server using [`Suspend-PodeServer`](../../../Functions/Core/Resume-PodeServer) and [`Resume-PodeServer`](../../../Functions/Core/Suspend-PodeServer).
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply docs/Tutorials/ServerOperation/Restarting/Overview.md:13]: copy and paste with no review :(

File: src/Private/Dump.ps1 [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+
+        Write-PodeHost -ForegroundColor Yellow 'Preparing Memory Dump ...'
+    }
+
+    # Process block to handle each pipeline input
+    process {
+        # Ensure Dump directory exists in the specified path
+        if ( $Path -match '^\.{1,2}([\\\/]|$)') {
+            $Path = [System.IO.Path]::Combine($PodeContext.Server.Root, $Path.Substring(2))
+        }
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:121]: There was a reason why Get-PodeRelativePath was not working the way a wanted. but I don't remember. Let me try again. I'll solve this converasation later

File: src/Private/Dump.ps1 [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+                ProcessName   = $process.ProcessName
+                WorkingSet    = [math]::Round($process.WorkingSet64 / 1MB, 2)
+                PrivateMemory = [math]::Round($process.PrivateMemorySize64 / 1MB, 2)
+                VirtualMemory = [math]::Round($process.VirtualMemorySize64 / 1MB, 2)
+            }
+        )
+
+        # Capture the code causing the exception
+        $scriptContext = @()
+        $exceptionDetails = @()
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:141]: you have a point

File: src/Private/Dump.ps1 [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+
+            $scriptContext += [Ordered]@{
+                ScriptName      = $ErrorRecord.InvocationInfo.ScriptName
+                Line            = $ErrorRecord.InvocationInfo.Line
+                PositionMessage = $ErrorRecord.InvocationInfo.PositionMessage
+            }
+
+            # Capture stack trace information if available
+            $stackTrace = if ($ErrorRecord.Exception.StackTrace) {
+                $ErrorRecord.Exception.StackTrace -split "`n"
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:154]: for readability. but let me try without

File: src/Private/Dump.ps1 [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+        }
+
+        # Collect variables by scope
+        $scopedVariables = Get-PodeDumpScopedVariable
+
+        # Check if RunspacePools is not null before iterating
+        $runspacePoolDetails = @()
+
+        # Retrieve all runspaces related to Pode ordered by name so the Main runspace are the first to be suspended (To avoid the process hunging)
+        $runspaces = Get-Runspace | Where-Object { $_.Name -like 'Pode_*' } | Sort-Object Name
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:175]: Copy and paste from Server.ps1 :)

File: src/Private/Dump.ps1 [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+
+        $runspaceDetails = @{}
+        foreach ($r in $runspaces) {
+            $runspaceDetails[$r.Name] = @{
+                Id                  = $r.Id
+                Name                = $r.Name
+                InitialSessionState = $r.InitialSessionState
+                RunspaceStateInfo   = $r.RunspaceStateInfo
+            }
+            $runspaceDetails[$r.Name].ScopedVariables = Get-PodeRunspaceVariablesViaDebugger -Runspace $r
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:185]: I'm thinking ......................................................................................................................................................................................................................................................
.
.
..................................................

ERROR NO ANSWER FOUND !!!!!!!!!!!!!!

File: src/Private/Dump.ps1 [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+            ExceptionDetails = $exceptionDetails
+            ScopedVariables  = $scopedVariables
+            RunspacePools    = $runspacePoolDetails
+            Runspace         = $runspaceDetails
+        }
+
+        # Determine file extension and save format based on selected Format
+        switch ($Format) {
+            'json' {
+                $dumpFilePath = Join-Path -Path $Path -ChildPath "PowerShellDump_$(Get-Date -Format 'yyyyMMdd_HHmmss').json"
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:226]: good point

File: src/Private/Dump.ps1 [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+        $result = @{}
+        try {
+            foreach ($key in $InputObject.Keys) {
+                try {
+                    # Serialize each key-value pair in the hashtable
+                    $strKey = $key.ToString()
+                    $result[$strKey] = ConvertTo-PodeSerializableObject -InputObject $InputObject[$key] -MaxDepth $MaxDepth -CurrentDepth ($CurrentDepth + 1)
+                }
+                catch {
+                    Write-PodeHost $_ -ForegroundColor Red
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:454]: The idea is to show on the console But I agree I completely forgot the Log

File: src/Private/Dump.ps1 (hunk 1 above)
Comment (mdaneri) [reply src/Private/Dump.ps1:503]: I never encountered this issue before. But your implementation is better

File: src/Private/Endpoints.ps1 [hunk 10]
```diff
@@ -385,4 +385,49 @@ function Get-PodeEndpointByName {
...
+    This command will output details of all endpoints the Pode server is currently
+    listening on, including their URLs and any relevant flags.
+
+.NOTES
+    This function uses `Write-PodeHost` to display messages, with the `Yellow` foreground
+    color for clarity. It ensures each endpoint is displayed with its associated flags,
+    enhancing visibility of specific configurations like `DualMode`.
+#>
+function Show-PodeEndPointConsoleInfo {
+    if ($PodeContext.Server.EndpointsInfo.Length -gt 0) {
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Endpoints.ps1:413]: I know you don't like indented code :) 
It's fault of my college days. The professor was obsessed with what in English is called Formal Method. Where any function has one entry point and one only exit. Otherwise the demonstration of the correctness was a nightmare. 
Anyway I haded it :D

File: src/Private/Server.ps1 [hunk 11]
```diff
@@ -147,77 +155,92 @@ function Start-PodeInternalServer {
...
+        $ShowHeader
+    )
+
+    if ($PodeContext.Server.Quiet) {
+        return
+    }
+
+    if ($ClearHost) {
+        Clear-Host
+    }
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Server.ps1:200]: sure

File: src/Private/Server.ps1 [hunk 12]
```diff
@@ -147,77 +155,92 @@ function Start-PodeInternalServer {
...
+        Write-PodeHost "    Ctrl+U   : $resumeOrSuspend" -ForegroundColor Cyan
+
+        if ((Get-PodeEndpointUrl) -and !($PodeContext.Server.Suspended)) {
+            Write-PodeHost "    Ctrl+B   : $($Podelocale.OpenHttpEndpointMessage)" -ForegroundColor Cyan
+        }
+
+        if ($PodeContext.Server.Debug.Dump.Enabled) {
+            Write-PodeHost "    Ctrl+D   : $($Podelocale.GenerateDiagnosticDumpMessage)" -ForegroundColor Cyan
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Server.ps1:236]: I like it. 
let me move these changes to another branch

File: src/Private/Server.ps1 [hunk 13]
```diff
@@ -335,6 +358,9 @@ function Restart-PodeInternalServer {
         Close-PodeDisposable -Disposable $PodeContext.Tokens.Restart
         $PodeContext.Tokens.Restart = [System.Threading.CancellationTokenSource]::new()
 
+        Close-PodeDisposable -Disposable $PodeContext.Tokens.Dump
+        $PodeContext.Tokens.Dump = [System.Threading.CancellationTokenSource]::new()
+
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Server.ps1:363]: Done

File: src/Private/Dump.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Dump.ps1:121]: there was a reason 

C:\Users\m_dan\Documents\GitHub\Pode\examples\.\Dump\PowerShellDump_20241130_084939.json

the .\

I'm going to fix Get-PodeRelativePath

File: docs/Getting-Started/Debug.md [hunk 14]
```diff
@@ -246,3 +253,93 @@ Add-PodeSchedule -Name 'TestSchedule' -Cron '@hourly' -ScriptBlock {
...
+            # Invoke a memory dump when a critical error occurs
+            $_ | Invoke-PodeDump -Halt
+        }
+    }
+}
+```
+
+In this setup, if an error occurs in the route, `Invoke-PodeDump` is called, capturing the current state and halting the application if the `-Halt` switch is set.
+
+### Dumping Options and Best Practices
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply docs/Getting-Started/Debug.md:336]: the file is huge. My dump is more than 1M lines

File: src/Public/Core.ps1 [hunk 15]
```diff
@@ -203,21 +206,51 @@ function Start-PodeServer {
             }
 
             # sit here waiting for termination/cancellation, or to restart the server
-            while (!(Test-PodeTerminationPressed -Key $key) -and !($PodeContext.Tokens.Cancellation.IsCancellationRequested)) {
+            while (  !($PodeContext.Tokens.Cancellation.IsCancellationRequested)) {
                 Start-Sleep -Seconds 1
 
-                # get the next key presses
-                $key = Get-PodeConsoleKey
+                if (!$PodeContext.Server.DisableTermination) {
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Public/Core.ps1:212]: I created a new $PodeContext.Server.DisableConsoleInput just to disable any keyboard input

File: src/Private/Helpers.ps1 [hunk 16]
```diff
@@ -3766,4 +3775,434 @@
...
+        # Set resume action to Continue
+        $resumeAction = [System.Management.Automation.DebuggerResumeAction]::Stop
+
+        # Create the DebuggerStopEventArgs
+        $eventArgs = [System.Management.Automation.DebuggerStopEventArgs]::new($null, $breakpoints, $resumeAction)
+
+        # Invoke the method
+        $methodInfo.Invoke($Runspace.Debugger, @($eventArgs))
+
+        Write-Host "DebuggerStopEvent raised successfully."
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Private/Helpers.ps1]: removed

File: src/Public/Core.ps1 [hunk 17]
```diff
@@ -203,21 +206,51 @@ function Start-PodeServer {
...
+                    $key = Get-PodeConsoleKey
+                }
 
                 # check for internal restart
                 if (($PodeContext.Tokens.Restart.IsCancellationRequested) -or (Test-PodeRestartPressed -Key $key)) {
+                    Clear-PodeKeyPressed
                     Restart-PodeInternalServer
                 }
 
+                if (($PodeContext.Tokens.Dump.IsCancellationRequested) -or (Test-PodeDumpPressed -Key $key) ) {
```
[the remark this answers is not in the document]
Comment (mdaneri) [reply src/Public/Core.ps1:248]: Let me remove the 'halt' part. I was not sure too about this implementation

File: docs/Getting-Started/Debug.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (mdaneri) [reply docs/Getting-Started/Debug.md:330]: Done
